text/profiles/0237-profiles-data-model.md

```diff
@@ -0,0 +1,1067 @@
+  // https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/README.md#attribute
+  // Attribute keys MUST be unique (it is not allowed to have more than one
+  // attribute with the same key).
+  repeated opentelemetry.proto.common.v1.KeyValue attributes = 1;
+
+  // dropped_attributes_count is the number of attributes that were discarded. Attributes
+  // can be discarded because their keys are too long or because there are too many
+  // attributes. If this value is 0, then no attributes were dropped.
+  uint32 dropped_attributes_count = 2;
+}
+
+// A stacktrace is a sequence of locations. Order of locations goes from callers to callees. Many stacktraces will point to the same locations. The link between stacktraces, attribute sets, links, values and timestamps is implicit and is based on the order of the elements in the corresponding tables in ProfileType message.
+message Stacktrace {
+  repeated uint32 location_indices = 1;
+}
+
+// AggregationTemporality defines how a profile aggregator reports aggregated
+// values. It describes how those values relate to the time interval over
+// which they are aggregated.
+enum AggregationTemporality {
+  AGGREGATION_TEMPORALITY_UNSPECIFIED = 0; // Temporality unspecified.
+  AGGREGATION_TEMPORALITY_DELTA = 1;       // Delta aggregation over time.
+  AGGREGATION_TEMPORALITY_CUMULATIVE = 2;  // Cumulative aggregation over time.
+}
+
+// Represents a single profile type. It implicitly creates a connection between Stacktraces, Links, AttributeSets, values and timestamps. The connection is based on the order of the elements in the corresponding lists. This implicit connection creates an ephemeral structure called Sample. The length of reference lists must be the same. It is acceptable however for timestamps, links and attribute set lists to be empty. It is not acceptable for stacktrace or values lists to be empty.
+message ProfileType {
+  // aggregation_temporality describes if the aggregator reports delta changes
+  // since last report time, or cumulative changes since a fixed start time.
+  AggregationTemporality aggregation_temporality = 1;
+
+  // Profiler sample rate in Hz. This parameter indicates the frequency at which samples are collected, specifically for CPU profiles. Common values are 99 or 100. [Optional].
+  uint64 sample_rate = 2;
+
+  // Index into the string table for the type of the sample. Example values are "cpu", "alloc_objects", "alloc_bytes", "block_contentions". Full list is defined in https://github.com/open-telemetry/semantic-conventions
+  uint32 type_index = 3;
+
+  // Index into the string table for the unit of the sample. Example values are "ms", "ns", "samples", "bytes". Full list is defined in https://github.com/open-telemetry/semantic-conventions
+  uint32 unit_index = 4;
+
+  // List of indices referring to Stacktraces in the Profile's stacktrace table.
+  repeated uint32 stacktrace_indices = 10;
```

Comment:
One of many cases where _index values are uint32. This is potentially painful in languages like Java that may deserialize to arrays with smaller index range (because signed ints), making some data inaccessible. Probably a non-issue, as the dominant use case is wire messages that will cap at 4MB total size anyhow, so odds of it ever happening are small. OTOH if the format is used for files too and you build one by aggregating many smaller files (common in pprof) then you potentially wind up with data that can be handled in some languages but not others.

Reply:
We work with stack trace datasets that may exceed a 32-bit index and require a 64-bit index, especially after aggregation (trillions vs billions). What's the story for huge datasets like this? Would it not be in this format directly?